lib/cli/run-helpers.js

```diff
@@ -0,0 +1,335 @@
+      if (typeof newFiles === 'string') {
+        newFiles = [newFiles];
+      }
+      newFiles = newFiles.filter(fileName =>
+        (argv.exclude || []).every(pattern => !minimatch(fileName, pattern))
+      );
+    }
+
+    files = files.concat(newFiles);
+  });
+
+  if (!files.length) {
+    console.error(chalk.red(`${symbols.error} No test files found`));
+    process.exit(1);
+  }
+
+  const fileArgs = (argv.file || []).map(filepath => path.resolve(filepath));
+  files = files.map(filepath => path.resolve(filepath));
+
+  // ensure we don't sort the stuff from fileArgs; order is important!
+  if (argv.sort) {
+    files.sort();
+  }
+
+  // add files given through --file to be ran first
+  files = fileArgs.concat(files);
+  debug('files (in order): ', files);
+  return files;
+};
+
+/**
+ * Actually run tests
+ * @param {Mocha} mocha - Mocha instance
+ * @param {yargs.Arguments} argv - Arguments
+ * @param {string[]} files - Array of test files
+ * @private
+ */
+exports.runMocha = (mocha, argv, files) => {
+  let runner;
+
+  if (argv.watch) {
+    console.log();
+    hideCursor();
+    process.on('SIGINT', () => {
+      showCursor();
+      console.log('\n');
+      process.exit(130);
+    });
+
+    const watchFiles = utils.files(cwd, argv.watchExtensions);
+    let runAgain = false;
+
+    const loadAndRun = () => {
+      try {
+        mocha.files = files;
+        runAgain = false;
+        runner = mocha.run(() => {
+          runner = null;
+          if (runAgain) {
+            rerun();
+          }
+        });
+      } catch (e) {
+        console.log(e.stack);
+      }
+    };
+
+    const purge = () => {
+      watchFiles.forEach(file => {
+        delete require.cache[file];
+      });
+    };
+
+    loadAndRun();
+
+    const rerun = () => {
+      purge();
+      stop();
+      if (!argv.grep) {
+        mocha.grep(null);
+      }
+      mocha.suite = mocha.suite.clone();
+      mocha.suite.ctx = new Mocha.Context();
+      mocha.ui(argv.ui);
+      loadAndRun();
+    };
+
+    utils.watch(watchFiles, () => {
+      runAgain = true;
+      if (runner) {
+        runner.abort();
+      } else {
+        rerun();
+      }
+    });
+  } else {
+    // load
+    mocha.files = files;
+    runner = mocha.run(argv.exit ? exit : exitLater);
+  }
+
+  process.on('SIGINT', () => {
+    runner.abort();
+
+    // This is a hack:
+    // Instead of `process.exit(130)`, set runner.failures to 130 (exit code for SIGINT)
+    // The amount of failures will be emitted as error code later
+    runner.failures = 130;
+  });
+};
+
+/**
+ * Used for `--reporter` and `--ui`.  Ensures there's only one, and asserts
+ * that it actually exists.
+ * @todo XXX This must get run after requires are processed, as it'll prevent
+ * interfaces from loading.
+ * @param {yargs.Arguments} argv - Arguments
+ * @param {string} option - Resolvable module name or path
+ * @param {Object} map - An object perhaps having key `option`
+ * @private
+ */
```

Comment:
This looks interesting. Is there a corresponding `registerPlugin`?

Reply:
Maybe "Plugin" is a misnomer.

A "Plugin" in this case is a module which, when `--require`'d, adds itself to `Mocha.interfaces` or `Mocha.reporter`.

We could provide a nice API to do that sort of thing, but there's already plenty of scope creep happening

Reply:
`mocha.use(myReporter)`

Reply:
let's deal w/ that later